### PR TITLE
AST representation changes

### PR DESCRIPTION
**Pull request by sonologico**
Opening the PR to continue the discussion started on #228.

Changes to the AST are:

* Remove the record wrapper for each node and inline the attribute field
* Make the block type parametric with the goal of using this flexibility to later have a way to parse and return the AST with source locations and whatever information we might want per node.

### CODE REVIEW

**Review by shonfeder — COMMENTED** [review]
I left a few suggestions, but they are just that. Let me know if you think any of them aren't for the better :)

I'm psyched to get this in place!

File: src/ast.ml
Comment (jfrolich) [reply src/ast.ml:27]: I think the only one that is confusing is the second string, perhaps we can alias a string type to a more descriptive name?

Comment (shonfeder) [reply src/ast.ml:27]: Yep, that's right, just the two `string`s. It's not clear which is which (tho, imo, it's fairly natural to assume the first is the label and the second the code, and it'll be immediately obvious in nearly every case once there is actually content, so during pattern matching).

I think adding a type alias will cause confusing noise with type inference, since it'll lead other string values to show up as whatever the alias is named.

After sleeping on this a bit I'm  inclined to think we don't need to worry about this much: we can document it, and it's various straight forward. Happy to leave it up to @sonologico to do whatever he thinks best here.

File: src/ast.ml
Comment (sonologico) [reply src/ast.ml:21]: Makes sense to me. I have made that change locally.

File: src/ast.ml
Comment (sonologico) [reply src/ast.ml:27]: I agree that aliasing has some unfortunate consequences. I'll think about it a little more, but I'm also inclined to leave it as is.

Comment (jfrolich) [reply src/ast.ml:27]: I've seen this in some other OCaml codebases. Didn't know about the unintended consequences. I think it's fine for now.

File: src/ast.ml
Comment (sonologico) [reply src/ast.ml:54]: That's a good point. Haven't added a comment yet. Not sure if I want to add a comment here or at the "external interface" level (`omd.mli`). I'm leaning towards adding it here. I'll add something before we merge this.

File: src/ast.ml
Comment (sonologico) [reply src/ast.ml]: I can see the benefit of it. I don't feel strongly about it, but I would either
* keep it as is and wait until we have a need for functors (which is what I'd favour)
* do it in the previous style in which the constructors are part of a functor so that they are also namespaced

**Comment by sonologico** [comment]
Let me know what you think after I have changed the rest of the code.

If we're going with this, I think that removing the `open Ast` line in some files and using the full path throughout the code would be better. What do you think?

File: src/block.ml
Comment (shonfeder) [inline src/block.ml:46]: I feel like the clarity gained by being able to remove this `mk` function thanks to the new Ast speaks the improvement here. :+1:

File: src/omd.mli
Comment (shonfeder) [inline src/omd.mli:49]: How about exposing the `Ast.t` in this interface and then using it here?

```suggestion
type doc = attributes Ast.t list
```

**Review by shonfeder — COMMENTED** [review]
Looking good to me! Just two small comments.

After having propagated the AST through the code base, what's your feeling on its merits compared to the current representation? Do you feel like it's an improvement to work with?

**Comment by shonfeder** [comment]
> If we're going with this, I think that removing the `open Ast` line in some files and using the full path throughout the code would be better. What do you think?

I'm fine with narrower scoping of the module if you think it's an improvement :+1:

**Comment by sonologico** [comment]
> Looking good to me! Just two small comments.
> 
> After having propagated the AST through the code base, what's your feeling on its merits compared to the current representation? Do you feel like it's an improvement to work with?

I think it's an improvement overall.

* I like having just the variants without the wrapping records. The small downside is that we might want to add a function to pattern match and return the attributes uniformly. I think it's worth it.
* Adding the type parameter for the attributes has made types a bit noisier. I think it's worth it, because it hasn't made the type harder to work with and it makes adding support for an alternative more detailed representation (with locations, etc) clearer.

File: src/omd.mli
Comment (sonologico) [reply src/omd.mli:49]: This is the thing that is making me consider moving back to using a functor as we were using before in the Ast module. There's no way to expose block without a parameter here. It feels weird to have `Omd.block`, ``Omd.t` and `Omd.doc`.

I'd like to either

1. Pick a different name for `block` and use the name `block` for `t` and update the `Ast` module to match it
2. Go back to having the functor in the `Ast` module as it was before so that `block` doesn't have a type parameter and we don't need a `t`.

Let me know what you think and I'll try to wrap up this up by tomorrow so that we can unblock the other things.

Comment (shonfeder) [reply src/omd.mli]: I'm afraid I don't quite follow. I think it makes sense to have a these two different type constructors exposed:

- `('attr, 'a) block`: types for block-level elements of any attributes and any content  
- `'attr Ast.t`: types for the standard markdown AST, over any attributes 

And to have this additional type:

- `doc = attributes t list`: the type for the standard representation of a markdown document, with the standard attributes. 

Tho I could also see an argument `Ast.t` being defines as a list of blocks.
(Well, if I were designing from scratch, I'd probably call `doc` `Omd.t`, but that's inessential.)

That said, if you think some use of simple functors will improve the interface and make a nicer presentation of types, I wouldn't be opposed. Do you want to sketch out what you have in mind?

Comment (sonologico) [reply src/omd.mli]: The `Ast` module used a [functor](https://github.com/ocaml/omd/blob/master/src/ast.ml#L30). I'm thinking of reintroducing that so that `Omd` can have the following signature:

```
(* ...  *)

type 'attr inline =
  | Concat of 'attr * 'attr inline list
  | Text of 'attr * string
  | Emph of 'attr * 'attr inline
  | Strong of 'attr * 'attr inline
  | Code of 'attr * string
  | Hard_break of 'attr
  | Soft_break of 'attr
  | Link of 'attr * 'attr inline link
  | Image of 'attr * 'attr inline link
  | Html of 'attr * string

(* ... *)

type 'attr block =
  | Paragraph of 'attr * 'attr inline
  | List of 'attr * list_type * list_spacing * 'attr block list list
  | Blockquote of 'attr * 'attr block list
  | Thematic_break of 'attr
  | Heading of 'attr * int * 'attr inline
  | Code_block of 'attr * string * string
  | Html_block of 'attr * string
  | Definition_list of 'attr * 'attr def_elt list

type doc = attributes block list
```

With the current setup in my branch, we can't really expose Ast.t in Omd without exposing the Ast module as well (maybe with a restricted signature).  I find that this Omd signature, the solution currently in my branch, is awkward:

```
(* ... *)
type ('attr, 'a) block =
  | Paragraph of 'attr * 'a
  | List of 'attr * list_type * list_spacing * ('attr, 'a) block list list
  | Blockquote of 'attr * ('attr, 'a) block list
  | Thematic_break of 'attr
  | Heading of 'attr * int * 'a
  | Code_block of 'attr * string * string
  | Html_block of 'attr * string
  | Definition_list of 'attr * 'a def_elt list

type 'attr t = ('attr, 'attr inline) block (* equivalent to Ast.t *)

type doc = attributes t list
(** A markdown document *)
```

I would either change the name of `block` to something else and change `t` to `block` or change `t` to something else.

File: src/ast.ml
Comment (sonologico) [reply src/ast.ml:61]: I have added a comment. Let me know if it's good enough.

File: src/omd.mli
Comment (sonologico) [reply src/omd.mli]: Added back the functor to the `Ast` module. There's no real cons internally besides a heavier weight syntax. I prefer the simpler external interface with a single parameter for the attributes on the `block` and `inline` types in `Omd`.

If you give me your approval, I'll squash the commits and merge this.

**Review by shonfeder — APPROVED** [review]
I'm not crazy about the functorial code here, but that's inherited :)

I suspect we can maybe find a nicer way of wrangling these types in the future, but I think the external API exposed in terms of the AST won't change!

Looks good to me! Thanks for the nice work here :)

**Pull request merged**
